akka-remote/src/main/scala/akka/remote/artery/TestStage.scala

```diff
@@ -131,19 +162,22 @@ private[remote] class InboundTestStage(inboundContext: InboundContext, state: Sh
 
         // InHandler
         override def onPush(): Unit = {
-          val env = grab(in)
-          env.association match {
-            case OptionVal.None ⇒
-              // unknown, handshake not completed
-              push(out, env)
-            case OptionVal.Some(association) ⇒
-              if (state.isBlackhole(inboundContext.localAddress.address, association.remoteAddress)) {
-                log.debug(
-                  "dropping inbound message [{}] from [{}] with UID [{}] because of blackhole",
-                  Logging.messageClassName(env.message), association.remoteAddress, env.originUid)
-                pull(in) // drop message
-              } else
+          if (state.shouldFailInbound) failStage(state.getInboundFailureOnce)
```

Comment:
to be strict this should be an atomic check/get/update, i.e. remove `shouldFailInbound` and return Option from `getInboundFailureOnce`